### PR TITLE
Add exclusion rules to filter unwanted events

### PR DESCRIPTION
**Pull request by acbox**

### CODE REVIEW

**Comment by hoerup** [comment]
could possibly solve #206 ?

**Comment by hoerup** [comment]
can the filtering also exclude CRD's ?
and thereby solve  #210

File: Tiltfile [hunk 1]
```diff
@@ -0,0 +1,57 @@
+#
```
Comment (nurland) [inline Tiltfile:1]: Can you please move this file out to a separate PR, we'd like to keep PRs focused on one feature, thanks!
It would also be great if we can update readme with instructions on how to run tile with sloop.

File: pkg/sloop/common/utilities.go [hunk 2]
```diff
@@ -37,3 +37,13 @@ func Contains(stringList []string, elem string) bool {
 func GetFilePath(filePath string, fileName string) string {
 	return path.Join(filePath, fileName)
 }
+
+func Truncate(text string, width int) (string, error) {
```
Comment (nurland) [inline pkg/sloop/common/utilities.go:41]: can you please add unit tests for Truncate?

File: pkg/sloop/ingress/kubewatcher_test.go [hunk 3]
```diff
@@ -81,21 +82,58 @@ func Test_bigPicture(t *testing.T) {
...
+	}
+
+	// await events
+	result1 := <-outChan
+	result2 := <-outChan
+	result3 := <-outChan
+
+	if !strings.Contains(result1.Payload, `"name":"ns"`) { t.Error("Expected event not found") }
+	if !strings.Contains(result2.Payload, `"name":"s1"`) { t.Error("Expected event not found") }
+	if !strings.Contains(result3.Payload, `"name":"s3"`) { t.Error("Expected event not found") }
```
Comment (nurland) [inline pkg/sloop/ingress/kubewatcher_test.go:136]: should we check that `"name":"s2"` is not present?

File: pkg/sloop/ingress/kubewatcher_test.go [hunk 4]
```diff
@@ -81,21 +82,58 @@ func Test_bigPicture(t *testing.T) {
 	masterURL := "url"
 	kubeContext := "" // empty string makes things work
 	enableGranularMetrics := true
-	kw, err := NewKubeWatcherSource(kubeClient, outChan, resync, includeCrds, time.Duration(10*time.Second), masterURL, kubeContext, enableGranularMetrics)
+	exclusionRules := map[string][]any{
```
Comment (nurland) [inline pkg/sloop/ingress/kubewatcher_test.go:85]: can we also add a test where exclusionRules are not provided, i.e. uses the default value of `ExclusionRules:           map[string][]any{},`

**Review by nurland — COMMENTED** [review]
Thank you @acbox for the contribution, this is great! Added some comments, but otherwise lgtm

File: pkg/sloop/ingress/kubewatcher_test.go (hunk 4 above)
Comment (acbox) [reply pkg/sloop/ingress/kubewatcher_test.go:133]: I've added seperate tests for default and non-default exclusion rules

File: Tiltfile (hunk 1 above)
Comment (acbox) [reply Tiltfile]: Removed, not sure this warrants a separate PR yet as there were some unresolved issues with the workflow incorporating Tilt. The Tiltfile also relied on running a local cluster with Kind, which is perhaps too opinionated.

File: pkg/sloop/common/utilities.go (hunk 2 above)
Comment (acbox) [reply pkg/sloop/common/utilities.go]: Added unit tests, doing this highlighted some bugs in Truncate that I've fixed.

File: pkg/sloop/ingress/kubewatcher_test.go (hunk 3 above)
Comment (acbox) [reply pkg/sloop/ingress/kubewatcher_test.go]: Could you give an example? I found this the most straightforward way. I changed the `if` statements to `assert` calls, bit cleaner.

File: pkg/sloop/ingress/kubewatcher.go [hunk 5]
```diff
@@ -360,6 +373,41 @@ func (i *kubeWatcherImpl) refreshCrdInformers(masterURL string, kubeContext stri
...
+func (i *kubeWatcherImpl) eventExcluded(kind string, resourceJson string) (bool) {
+	filters := i.getExclusionRules(kind)
+	for _, logic := range filters {
+		logicJson, err := json.Marshal(logic)
+		if err != nil {
+			glog.Errorf(`Failed to parse event filtering rule "%s": %s`, logic, err)
+			return false
+		}
+		var result bytes.Buffer
+		err = jsonlogic.Apply(
```
Comment (nurland) [inline pkg/sloop/ingress/kubewatcher.go:396]: please add err check

Comment (acbox) [reply pkg/sloop/ingress/kubewatcher.go:396]: Done

**Review by nurland — APPROVED** [review]
Thank you for your contribution! LGTM

**Comment by acbox** [comment]
Pushed a fix for the [failed](https://github.com/salesforce/sloop/actions/runs/4927800936/jobs/8805861492) GitHub Action test and signed the CLA.

**Review by nurland — APPROVED** [review]

**Comment by acbox** [comment]
This PR should pass the CI tests now as per the commit comment here https://github.com/salesforce/sloop/pull/259/commits/ae0b4d0a403078f4cad62948bda67746d8ac3ded

**Comment by sana-jawad** [comment]
@acbox thanks for your contribution.
